volatility3/framework/plugins/linux/netfilter.py

```diff
@@ -0,0 +1,716 @@
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+from dataclasses import dataclass, field
+from abc import ABC, abstractmethod
+import logging
+
+from typing import Iterator, List, Tuple
+from volatility3.framework import (
+    class_subclasses,
+    constants,
+    interfaces,
+    renderers,
+)
+from volatility3.framework.renderers import format_hints
+from volatility3.framework.configuration import requirements
+from volatility3.framework.symbols import linux
+from volatility3.plugins.linux import lsmod
+
+vollog = logging.getLogger(__name__)
+
+
+@dataclass
+class Proto:
+    name: str
+    hooks: Tuple[str] = field(default_factory=tuple)
+
+
+PROTO_NOT_IMPLEMENTED = Proto(name="UNSPEC")
+
+NF_INET_HOOKS = ("PRE_ROUTING", "LOCAL_IN", "FORWARD", "LOCAL_OUT", "POST_ROUTING")
+NF_DEC_HOOKS = (
+    "PRE_ROUTING",
+    "LOCAL_IN",
+    "FORWARD",
+    "LOCAL_OUT",
+    "POST_ROUTING",
+    "HELLO",
+    "ROUTE",
+)
+NF_ARP_HOOKS = ("IN", "OUT", "FORWARD")
+NF_NETDEV_HOOKS = ("INGRESS", "EGRESS")
+LARGEST_HOOK_NUMBER = max(
+    len(NF_INET_HOOKS), len(NF_DEC_HOOKS), len(NF_ARP_HOOKS), len(NF_NETDEV_HOOKS)
+)
+
+
+class AbstractNetfilter(ABC):
+    """Netfilter Abstract Base Classes handling details across various
+    Netfilter implementations, including constants, helpers, and common
+    routines.
+    """
+
+    PROTO_HOOKS = (
+        PROTO_NOT_IMPLEMENTED,  # NFPROTO_UNSPEC
+        Proto(name="INET", hooks=NF_INET_HOOKS),  # From kernels 3.14
+        Proto(name="IPV4", hooks=NF_INET_HOOKS),
+        Proto(name="ARP", hooks=NF_ARP_HOOKS),
+        PROTO_NOT_IMPLEMENTED,
+        Proto(name="NETDEV", hooks=NF_NETDEV_HOOKS),
+        PROTO_NOT_IMPLEMENTED,
+        Proto(name="BRIDGE", hooks=NF_INET_HOOKS),
+        PROTO_NOT_IMPLEMENTED,
+        PROTO_NOT_IMPLEMENTED,
+        Proto(name="IPV6", hooks=NF_INET_HOOKS),
+        PROTO_NOT_IMPLEMENTED,
+        Proto(name="DECNET", hooks=NF_DEC_HOOKS),  # Removed in kernel 6.1
+    )
+    NF_MAX_HOOKS = LARGEST_HOOK_NUMBER + 1
+
+    def __init__(
+        self,
+        context: interfaces.context.ContextInterface,
+        config: interfaces.configuration.HierarchicalDict,
+    ):
+        self._context = context
+        self._config = config
+        symbol_table = self._config["kernel"]
+        self.vmlinux = context.modules[symbol_table]
+        self.layer_name = self.vmlinux.layer_name
+
+        modules = lsmod.Lsmod.list_modules(context, symbol_table)
+        self.handlers = linux.LinuxUtilities.generate_kernel_handler_info(
+            context, symbol_table, modules
+        )
+
+        self._set_data_sizes()
+
+    def _set_data_sizes(self):
+        self.ptr_size = self.vmlinux.get_type("pointer").size
+        self.list_head_size = self.vmlinux.get_type("list_head").size
+
+    @classmethod
+    def run_all(
+        cls,
+        context: interfaces.context.ContextInterface,
+        config: interfaces.configuration.HierarchicalDict,
```

Comment:
It looks like `config['kernel']` is the only time this is ever used? I'd be much happier passing through a module_name than a whole hierarchical dict? Was the expectation this might need extending or is there another good reason to pass config through?

Reply:
good catch